### PR TITLE
test: use address_to_scriptpubkey instead of RPC call

### PR DESCRIPTION
**Pull request by ismaelsadeeq**
PR #27269 enables the function address_to_scriptpubkey() to decode all address types and return their corresponding scriptpubkeys. As a result, there is no longer any need to call getaddressinfo or validateaddress RPCs in order to retrieve an address scriptpubkey, as explained in the comments on this pull request (see https://github.com/bitcoin/bitcoin/pull/27269#pullrequestreview-1353681933 and https://github.com/bitcoin/bitcoin/pull/27269#issuecomment-1481016118).

Instead of using RPC calls, this update replaces the process of obtaining an address scriptPubkey with the address_to_scriptpubkey method, resulting in improved performance for functional tests.

### CODE REVIEW

**Comment by ismaelsadeeq** [comment]
@josibake  @theStack @willcl-ark this is PR https://github.com/bitcoin/bitcoin/pull/27269 follow up, I will appreciate your reviews.
Thank you

File: test/functional/feature_nulldummy.py [hunk 1]
```diff
@@ -77,7 +78,8 @@ def run_test(self):
         cms = self.nodes[0].createmultisig(1, [self.pubkey])
         wms = self.nodes[0].createmultisig(1, [self.pubkey], 'p2sh-segwit')
         self.ms_address = cms["address"]
-        ms_unlock_details = {"scriptPubKey": self.nodes[0].validateaddress(self.ms_address)["scriptPubKey"],
+        spk = address_to_scriptpubkey(self.ms_address).hex()
+        ms_unlock_details = {"scriptPubKey":spk,
```
Comment (josibake) [inline test/functional/feature_nulldummy.py:82]: I'm not sure adding `spk` as a variable really helps anything here and it does add an extra line to be reviewed. For a refactor, I'd suggest keeping the diff as small and clear as possible

```suggestion
        ms_unlock_details = {"scriptPubKey": address_to_scriptpubkey(self.ms_address).hex(),
```

**Review by josibake — COMMENTED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/27349/commits/c73a0658c2c5c43d4ba2d8dbf2af71abe84b40e8

Nice to see a standard function for getting the `scriptPubKey` instead of a mix of RPC calls. Removing the RPC call _should_ make the tests run faster, but considering it wasn't used in many places, I doubt it makes a material difference. Left two non-blocking nits.

I also noticed the CI was failing, but when I rebased on master and ran the failing functional test locally, everything passed.

File: test/functional/test_framework/wallet.py [hunk 2]
```diff
@@ -101,7 +97,7 @@ def __init__(self, test_node, *, mode=MiniWalletMode.ADDRESS_OP_TRUE):
             self._scriptPubKey = key_to_p2pk_script(pub_key.get_bytes())
         elif mode == MiniWalletMode.ADDRESS_OP_TRUE:
             self._address, self._internal_key = create_deterministic_address_bcrt1_p2tr_op_true()
-            self._scriptPubKey = bytes.fromhex(self._test_node.validateaddress(self._address)['scriptPubKey'])
+            self._scriptPubKey = address_to_scriptpubkey(self._address)
```
[the remark this answers is not in the document]
Comment (ismaelsadeeq) [reply test/functional/test_framework/wallet.py:100]: Yes.

File: test/functional/feature_nulldummy.py (hunk 1 above)
Comment (ismaelsadeeq) [reply test/functional/feature_nulldummy.py:82]: Thanks.

**Comment by ismaelsadeeq** [comment]
> @ismaelsadeeq seems ready for review to me
Ready for review now.

**Comment by ismaelsadeeq** [comment]
> ACK [c73a065](https://github.com/bitcoin/bitcoin/commit/c73a0658c2c5c43d4ba2d8dbf2af71abe84b40e8)
> 
> Nice to see a standard function for getting the `scriptPubKey` instead of a mix of RPC calls. Removing the RPC call _should_ make the tests run faster, but considering it wasn't used in many places, I doubt it makes a material difference. Left two non-blocking nits.
> 
> I also noticed the CI was failing, but when I rebased on master and ran the failing functional test locally, everything passed.

Thank you for the review Josi
I rebased to master recently but it failed again.

**Comment by josibake** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/27349/commits/01e0a1679fb1deeae4daf674730de2e1a270139e

thanks for taking the suggestions! CI failure seems unrelated from what I can tell :shrug:

File: test/functional/feature_nulldummy.py [hunk 3]
```diff
@@ -77,7 +78,7 @@ def run_test(self):
         cms = self.nodes[0].createmultisig(1, [self.pubkey])
         wms = self.nodes[0].createmultisig(1, [self.pubkey], 'p2sh-segwit')
         self.ms_address = cms["address"]
-        ms_unlock_details = {"scriptPubKey": self.nodes[0].validateaddress(self.ms_address)["scriptPubKey"],
+        ms_unlock_details = {"scriptPubKey":address_to_scriptpubkey(self.ms_address).hex(),
```
Comment (josibake) [inline test/functional/feature_nulldummy.py:81]: nit: should be a space here

```suggestion
        ms_unlock_details = {"scriptPubKey": address_to_scriptpubkey(self.ms_address).hex(),
```

Comment (ismaelsadeeq) [reply test/functional/feature_nulldummy.py:81]: Thanks

File: test/functional/test_framework/address.py [hunk 4]
```diff
@@ -21,11 +21,15 @@
...
+from test_framework.script_util import (
+    keyhash_to_p2pkh_script,
+    program_to_witness_script,
+    scripthash_to_p2sh_script,
+)
 from test_framework.segwit_addr import (
     decode_segwit_address,
     encode_segwit_address,
 )
-
```
Comment (theStack) [inline test/functional/test_framework/address.py:28]: stylistic nit: we usually keep two blank lines after imports, so should rather add another one than remove one here (see also https://peps.python.org/pep-0008/#blank-lines).

**Review by theStack — COMMENTED** [review]
Concept ACK, thanks for following up!
LGTM, left just one small suggestion below.

File: test/functional/test_framework/address.py (hunk 4 above)
Comment (ismaelsadeeq) [reply test/functional/test_framework/address.py:28]: Thanks :100:

**Comment by josibake** [comment]
re-ACK https://github.com/bitcoin/bitcoin/pull/27349/commits/e47ce42f670fc43859c157766b342509ab5916f9

**Review by theStack — APPROVED** [review]
ACK e47ce42f670fc43859c157766b342509ab5916f9 🌱

**Pull request merged**